### DISCUSSION TITLE
Are people using the grid editor?

### DISCUSSION THREAD

**Discussion by john-science** [terrapower/armi#1722] [Polls]
> Results of the poll: 100% yes!

The grid editor feature (found by running `armi grids` on the command line) is in need of some updates:

1. It only supports Windows and MacOS, because it uses `wxpython`.
2. It is not mentioned at all in the ARMI docs.

But before I fix these issues, I want to make sure the feature is useful in ARMI.

**Comment by john-science** [comment]
@sombrereau @drewejohnson @ntouran @albeanth @opotowsky Just FYI.

**Reply by albeanth** [reply]
> Are people using the grid editor?

I'm not. Until a few hours ago I didn't even know it existed. 

> Do you think the grid editor is an important ARMI feature?

For my own use and general internal TP practice/use, I think no, it's not. 🤷 

_But_, I see validity in it conceptually. Being able to visualize core maps for assembly-based reactor designs _is_ indeed useful. 

So my answer to your poll is wishy-washy. No and yes. Sorry!

**Comment by jakehader** [comment]
Probably one of the most important features of ARMI!

**Reply by john-science** [reply]
Cool.  How often have you used it in the past, say, year?

**Reply by jakehader** [reply]
Well, given that I don't do so much analysis now-a-days how much I've used it in the past year may not be useful, but basically the grid editor makes it very easy to assign assemblies to grid locations for creating the core map and fuel management schemes for fuel shuffling in equilibrium analyses and for ASCII inputs in the blueprints. It's not impossible to define this by hand, but I'd say that doing this without a GUI is pretty error prone and easy to make mistakes. This is one feature of ARMI that I don't think we could do without and I'd be sad to see it go. It's unfortunate that wxPython is not cross platform compatible and maybe my exploratory GUI rewrite could be what supercedes this. 

When I was doing analysis work, I would use this at least weekly, if not daily. This feature is what originally wowed me when I joined TerraPower - credit to @ntouran

**Reply by albeanth** [reply]
I agree with Jake that it's a very useful tool to have. Editing assembly-based core maps by hand is.... a real pain. _Especially_ for hex-based reactors. Having a native capability in ARMI to visualize and be able to edit them is great. I think it would be a bit shortsighted to just rip it out.

That said though, we should be able to (or even _need_ to) have cross-platform/OS support. What if we compromise and deprecate it? That way we can dedicate resources to a new and improved revision that is cross-platform, modern, tested, etc.

**Comment by ntouran** [comment]
I agree that it's a 100% essential feature of ARMI. It is used every single day by people who are doing new core designs and reload analysis. For anyone focused on PSAR of a fresh core, they won't use it much for a few years (because the core layout is locked in), but once things get back into doing reload optimization and designing upgraded cores, this will be a every day feature. It's the only good way to specify core loading patterns. 

It is mentioned in the docs:
* here:  https://terrapower.github.io/armi/user/inputs.html#the-assembly-clicker (but link is broken!?)
* here: https://terrapower.github.io/armi/.apidocs/armi.utils.gridEditor.html (but render from apidocs is broken!?, should be rendering docstring from https://github.com/terrapower/armi/blob/a7a868b6668569678ed02e55def13fc185537173/armi/utils/gridEditor.py#L14-L30

Note also that it most definitely works on linux and any other POSIX system: https://github.com/terrapower/armi/issues/1721#issuecomment-2169128357

![image](https://github.com/terrapower/armi/assets/53948397/4039e8a9-9b6c-4439-92ab-5c558f9d7733)

The only action here should be to add more screenshots and instructions of how to use the grid editor to the docs.  UPDATE: I made a PR to do so: https://github.com/terrapower/armi/pull/1731

**Comment by slee-USNC** [comment]
I rely heavily on `grids` entry point. I cannot imagine myself rendering hexagonal periodic geometry by hand. It gets a bit hacky when I need to render non-core or sub-block grids, but it is still very useful. Hopefully we can switch back and forth between grids given a BP with multiple grids?

**Comment by john-science** [comment]
Okay, the vote are in and they are all "Yes". Great!

Thanks, all!

Nick has already kicked the ball of with documenting this feature, now I'll just have to decide if it needs more documentation.  (And it is currently untestable, so that needs fixing.)
